File: backend/api/assets/starters/rust-linkedlist/spec/linked_list.rs
// Spec skeleton for students — do not provide the implementation here.
// Fill in the node structure, fields, and methods as per the assignment.

#[derive(Clone, Debug)]
pub struct LinkedList<T> {
    // TODO: define your storage (e.g., head pointer, length)
    head: Link<T>,
    len: usize,
}

type Link<T> = Option<Box<Node<T>>>;

#[derive(Clone, Debug)]
struct Node<T> {
    // TODO: node fields
    elem: T,
    next: Link<T>,
}

impl<T> LinkedList<T> {
    pub fn new() -> Self {
        // TODO: initialize an empty list
        Self { head: None, len: 0 }
    }

    pub fn len(&self) -> usize {
        // TODO
        self.len
    }

    pub fn is_empty(&self) -> bool {
        // TODO
        self.len == 0
    }

    pub fn push_front(&mut self, _elem: T) {
        // TODO
        unimplemented!("push_front")
    }

    pub fn push_back(&mut self, _elem: T) {
        // TODO
        unimplemented!("push_back")
    }

    pub fn pop_front(&mut self) -> Option<T> {
        // TODO
        unimplemented!("pop_front")
    }

    pub fn clear(&mut self) {
        // TODO
        unimplemented!("clear")
    }

    pub fn front(&self) -> Option<&T> {
        // TODO
        unimplemented!("front")
    }

    pub fn back(&self) -> Option<&T> {
        // TODO
        unimplemented!("back")
    }

    pub fn insert_at(&mut self, _idx: usize, _elem: T) -> Result<(), &'static str> {
        // TODO
        unimplemented!("insert_at")
    }

    pub fn remove_at(&mut self, _idx: usize) -> Result<T, &'static str> {
        // TODO
        unimplemented!("remove_at")
    }

    pub fn to_vec(&self) -> Vec<&T> {
        // TODO: collect references to elements in order
        unimplemented!("to_vec")
    }

    pub fn from_slice(_slice: &[T]) -> Self
    where
        T: Clone,
    {
        // Optional helper
        unimplemented!("from_slice")
    }
}

File: backend/api/assets/starters/go-linkedlist/spec/linked_list.go
package main

// Spec skeleton (students implement these methods)

type node struct {
    val  int
    next *node
}

type LinkedList struct {
    head *node
    tail *node
    size int
}

func New() *LinkedList { return &LinkedList{} }
func (l *LinkedList) Len() int { return l.size }
func (l *LinkedList) IsEmpty() bool { return l.size == 0 }

func (l *LinkedList) Clear() { panic("TODO: Clear") }
func (l *LinkedList) PushFront(v int) { panic("TODO: PushFront") }
func (l *LinkedList) PushBack(v int) { panic("TODO: PushBack") }
func (l *LinkedList) PopFront() (bool, int) { panic("TODO: PopFront") }
func (l *LinkedList) Front() (int, bool) { panic("TODO: Front") }
func (l *LinkedList) Back() (int, bool) { panic("TODO: Back") }
func (l *LinkedList) InsertAt(idx int, v int) bool { panic("TODO: InsertAt") }
func (l *LinkedList) RemoveAt(idx int) bool { panic("TODO: RemoveAt") }
func (l *LinkedList) ToSlice() []int { panic("TODO: ToSlice") }

func (l *LinkedList) Copy() *LinkedList { panic("TODO: Copy") }
func MoveFrom(src *LinkedList) *LinkedList { panic("TODO: MoveFrom") }
func (l *LinkedList) MoveAssignFrom(src *LinkedList) { panic("TODO: MoveAssignFrom") }

